### PR TITLE
Enable building with OpenCV 3 & 4, along with some warning fixes

### LINKED ISSUE #60: Support for Ubuntu 20.04 LTS and OpenCV v4.2
With Ubuntu 20.04 LTS now shipping with OpenCV v4.2 , it would be nice update Kimera-VIO to support the latest long term support release of Ubuntu Focal and ROS releases Noetic and Foxy:

https://packages.ubuntu.com/focal/libopencv-dev

This should just evolve updating Kimera-VIO to transition it away from APIs that where deprecated in OpenCV v3 and removed in v4. From an initial look over, it seems that some of the c type headers will have to be explicitly imported, and switching to use the proper enum types for function arguments.

### PR DESCRIPTION
**Pull request by Wallbraker**
These changes enables me to build Kimera-VIO on Ubuntu 20.04 using the shipped OpenCV. The changes are done in such a way that it is still possible to build with `OpenCV 3.3.1`. In the other PR for this I saw some regressions in testing which is why I left `OpenCV 3.3.1` the one used in the Dockerfiles. First commit slightly based on changes by @ruffsl.

Fixes #60
Alternative to #61

### CODE REVIEW

**Comment by Wallbraker** [comment]
I have no Idea what's going on with the CI

**Comment by Wallbraker** [comment]
See https://github.com/borglab/gtsam/issues/398

**Comment by Wallbraker** [comment]
The issue (borglab/gtsam#398) should now be fixed, pushed a extra commit to re-trigger the CI.

**Comment by Wallbraker** [comment]
Is there anything I should change on the PR?

File: src/frontend/StereoFrame.cpp [hunk 1]
```diff
@@ -1209,8 +1209,8 @@ std::pair<StatusKeypointCV, double> StereoFrame::findMatchingKeypointRectified(
 
   // Refine keypoint with subpixel accuracy.
   if (sparse_stereo_params_.subpixel_refinement_) {
-    static const cv::TermCriteria criteria(CV_TERMCRIT_EPS + CV_TERMCRIT_ITER,
-                                           40, 0.001);
+    static const cv::TermCriteria criteria(
+        cv::TermCriteria::EPS + cv::TermCriteria::COUNT, 40, 0.001);
```
Comment (ToniRV) [inline src/frontend/StereoFrame.cpp:1213]: no preprocessor checks here?

File: src/frontend/feature-detector/FeatureDetectorParams.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ void SubPixelCornerFinderParams::print() const {
 
 bool SubPixelCornerFinderParams::parseYAML(const std::string& filepath) {
   YamlParser yaml_parser(filepath);
-  term_criteria_.type = CV_TERMCRIT_EPS + CV_TERMCRIT_ITER;
+  term_criteria_.type = cv::TermCriteria::EPS + cv::TermCriteria::COUNT;
```
Comment (ToniRV) [inline src/frontend/feature-detector/FeatureDetectorParams.cpp:44]: no preprocessor checks here?

File: src/frontend/feature-detector/NonMaximumSuppression.cpp [hunk 3]
```diff
@@ -51,7 +51,7 @@ std::vector<cv::KeyPoint> AdaptiveNonMaximumSuppression::suppressNonMax(
   }
   std::vector<int> Indx(responseVector.size());
   std::iota(std::begin(Indx), std::end(Indx), 0);
-  cv::sortIdx(responseVector, Indx, CV_SORT_DESCENDING);
+  cv::sortIdx(responseVector, Indx, cv::SortFlags::SORT_DESCENDING);
```
Comment (ToniRV) [inline src/frontend/feature-detector/NonMaximumSuppression.cpp:54]: no preprocessor checks here?

**Comment by ToniRV** [comment]
LGTM, just some questions.

**Comment by ToniRV** [comment]
Thank you, that's a great PR! I will increase the OpenCV version in 18.04 (and keep perhaps the same in 16.04) so that we test both versions without making Jenkins take too long.

**Comment by ToniRV** [comment]
Actually, can you modify Dockerfile_18_04 and add the changes here?

https://github.com/MIT-SPARK/Kimera-VIO/pull/61/files#diff-c9683af8e8ddc02e49349208a5717685

This way we test the compatibility with both OpenCV versions (I fear some unit test might break with the new version).

File: include/kimera-vio/loopclosure/LoopClosureDetector.h [hunk 4]
```diff
@@ -14,20 +14,17 @@
...
 #include <utility>
 #include <vector>
 
-#include <gtsam/geometry/Pose3.h>
-#include <gtsam/linear/NoiseModel.h>
-
-#include <opencv/cv.hpp>
-#include <opencv2/features2d.hpp>
-
-#include <DBoW2/DBoW2.h>
```
[the remark this answers is not in the document]
Comment (Wallbraker) [reply include/kimera-vio/loopclosure/LoopClosureDetector.h:29]: Those are not from me, I ran `clang-format` on my change and it changed it around. It's policy you put in place I just wanted to follow it.

File: src/frontend/StereoFrame.cpp (hunk 1 above)
Comment (Wallbraker) [reply src/frontend/StereoFrame.cpp:1213]: Those enums are already in OpenCV3 so no preprocessor checks needed.

File: src/frontend/feature-detector/FeatureDetectorParams.cpp (hunk 2 above)
Comment (Wallbraker) [reply src/frontend/feature-detector/FeatureDetectorParams.cpp:44]: Ditto

File: src/frontend/feature-detector/NonMaximumSuppression.cpp (hunk 3 above)
Comment (Wallbraker) [reply src/frontend/feature-detector/NonMaximumSuppression.cpp:54]: Ditto

**Comment by Wallbraker** [comment]
> Thank you, that's a great PR! I will increase the OpenCV version in 18.04 (and keep perhaps the same in 16.04) so that we test both versions without making Jenkins take too long.

No problems.

> Actually, can you modify Dockerfile_18_04 and add the changes here?
> 
> https://github.com/MIT-SPARK/Kimera-VIO/pull/61/files#diff-c9683af8e8ddc02e49349208a5717685
> 
> This way we test the compatibility with both OpenCV versions (I fear some unit test might break with the new version).

Looking at the other PR some tests will break most likely, If the CI breaks because of the failed tests can the PR even be merged? I would rather that be done in another PR after they have been investigated.

**Comment by Wallbraker** [comment]
Ping? @ToniRV

I have updated my changes to the latest master, the CI fault seems to be on master as well.

**Comment by Wallbraker** [comment]
@ToniRV Let me know if there is anything you want changed, as I said, I don't want to break the CI unnecessary. Maybe I could add a `Dockerfile` specifically for OpenCV 4 so that it is at least built?

**Comment by Wallbraker** [comment]
Ping @ToniRV 

I have updated the changes to master. Again if there is anything you want me to change please let me know. I would very much like to have this in before master moves again and I have to spend hours trying to figure out weird build breakages.

Cheers, Jakob.

**Comment by ToniRV** [comment]
Thank you @Wallbraker

**Pull request merged**